### ISSUE TITLE
BUG: tomopy.recon center array does not work properly

### ISSUE THREAD

**Issue by malte-storm** [tomopy/tomopy#450]
Hello,

passing an array with different centers for different slices (for center finding) does not work as expected. 
The center does not get updated for every slice if I use a single call with a list of slices and an array of centers. I tried to chase the problem through the code and as far as I can tell, the "center" is correct up to the call of "algorithm" in "_dist_recon" (in tomopy.recon.algorithm)

This is reproducible with different input datasets.

**To Reproduce (if applicable)**
```python
import h5py
import numpy as np
import tomopy
import skimage.io
import os
os.chdir('/dls/science/groups/i13/Malte/tomopy_recCenter')

data = np.load('./phantom_sino.npy')

theta = np.linspace(0, np.pi, num = 500, endpoint = False)
_centerarr = np.linspace(200, 300, num = 51)

tmp = np.tile(data, _centerarr.size)
tmp.shape = (data.shape[0], _centerarr.size, data.shape[1])

rec = tomopy.recon(tmp, theta, center=_centerarr, algorithm = 'gridrec')
skimage.io.imsave('./reco_batch_phantom.tif', rec.astype(np.float32))

rec2 = np.zeros((_centerarr.size, data.shape[1], data.shape[1]))
for i1 in range(_centerarr.size):
    rec2[i1] = tomopy.recon(data[:, np.newaxis,:], theta, center=_centerarr[i1], algorithm = 'gridrec')
skimage.io.imsave('./reco_single_phantom.tif', rec2.astype(np.float32))
```

**Expected behavior (if applicable)**
I would expect the results "rec" and "rec2" to be exactly the same (apart from variations due to limited precision floating point operations)

**Helpful images (if applicable)**
I did not upload any images or data as the problem seems to be reproducible with any kind of data.

**Platform Information:**
 - OS: Linux, RedHat 7
 - Python Version 3.7
 - TomoPy Version 1.4.0

**Additional context**

**Comment by malte-storm** [comment]
After some additional testing: The bug must be in the algorithm implementation of gridrec because it works fine if I change the chunking to nchunk=1 or with other algorithms (art, fbp)

**Comment by carterbox** [comment]
I agree that the centers are being passed to the c-functions with correct slicing. The centers seem to  match the slices being passed to each call to gridrec.

https://github.com/tomopy/tomopy/blob/139f4e82beea6b6dbdf54b9f531215d42dc59b5b/source/tomopy/recon/algorithm.py#L388-L396

The claim is that the variability of the reconstructions is caused by a difference in how the centers are handled. However, I'm not convinced that this is the case. This is because repeated calls to gridrec using the same parameters are only the same up to 2 decimal places.

```python
import numpy as np
import tomopy

# Create a test case from peppers that has 8 slices
theta = tomopy.angles(128)
data = tomopy.project(np.tile(tomopy.peppers(64), (8, 1, 1)), theta, pad=False)
_centerarr = np.arange(data.shape[2] // 2 - 4, data.shape[2] // 2 + 4)
assert _centerarr.size == data.shape[1]

print(f"data shape is {data.shape}")

# reconstruct using multiple centers
recb1 = tomopy.recon(
    tomo=data,
    theta=theta,
    center=_centerarr,
    algorithm='gridrec',
    sinogram_order=False,
    ncore=4,
)

# reconstruct using same parameters as above
recb4 = tomopy.recon(
    tomo=data,
    theta=theta,
    center=_centerarr,
    algorithm='gridrec',
    sinogram_order=False,
    ncore=4,
)

# This test will fail
np.testing.assert_allclose(recb1, recb4, atol=1e-3)
```

**Comment by carterbox** [comment]
@malte-storm Do you have any reason to believe that the **wrong** center is being used for some all slices? Or it it just that the reconstructions are not consistent within a small enough range of precision?

**Comment by malte-storm** [comment]
Yes, I believe the wrong centers are used. In your example, testing one set of recons for consistency yields:

```python
np.testing.assert_allclose(recb1[0], recb1[1], atol=1e-3)
Mismatch: 5.69%
np.testing.assert_allclose(recb1[1], recb1[2], atol=1e-3)
Mismatch: 96.9%
np.testing.assert_allclose(recb1[2], recb1[3], atol=1e-3)
Mismatch: 4.05%
np.testing.assert_allclose(recb1[3], recb1[4], atol=1e-3)
Mismatch: 96.9%
```
And if one looks at the actual images (e.g. 
![rec1b_comparison] (https://user-images.githubusercontent.com/10544058/65494471-d2917500-deac-11e9-853e-103157e82a64.png), there are some differences in the corners but the rest of the image is "similar" for pairs whereas I would expect similar diffrences between every pair of images.

I also attached one example with phantom data where the reconstruction center is shifted by two pixels for every image (multitiff) but if you look at it (reco_batch_phantom, multitiff image, had to zip it to attach), it is not updated properly (51 images distributed across 24 cores). But with nchunk=1, the reconstruction center is updated properly.


[reco_batch_phantom.zip](https://github.com/tomopy/tomopy/files/3646094/reco_batch_phantom.zip)
[reco_batch_chunk1_phantom.zip](https://github.com/tomopy/tomopy/files/3646097/reco_batch_chunk1_phantom.zip)

**Comment by carterbox** [comment]
The problem is that, slices are processed in pairs using the same center. Thus, the centers will be wrong for odd numbered slices any time a volume is processed in bulk.

To illustrate this quirk, let's use the following example where a different center is used every third index.

```python
import numpy as np
import tomopy
import matplotlib.pyplot as plt

# Create a test case from peppers that has 8 slices
theta = tomopy.angles(256)
data = tomopy.project(np.tile(tomopy.peppers(256), (8, 1, 1)), theta, pad=False)

# Make a sequence of centers
right, wrong = 256 / 2, 64
_centerarr = np.array([wrong, right, right] * 4)

# Bulk reconstruct using multiple centers
recb1 = tomopy.recon(
    tomo=data,
    theta=theta,
    center=_centerarr,
    algorithm='gridrec',
    sinogram_order=False,
    ncore=1,
)

# Plot these reconstructions in a row
plt.figure(dpi=200, figsize=(8, 1))
for i, s in enumerate(recb1):
    plt.subplot(1, 8, i + 1)
    plt.imshow(s, vmin=0, vmax=256)
plt.show()
```

![Figure_1](https://user-images.githubusercontent.com/9604511/65550053-63377780-dee4-11e9-850e-1ee5bd6ec9d5.png)

The reconstructed images clearly show that only the even indexed centers are being used, and the relevant code segment is shown below:

https://github.com/tomopy/tomopy/blob/139f4e82beea6b6dbdf54b9f531215d42dc59b5b/source/c/gridrec.c#L192-L206 

Looking into the Git history, it appears that reconstructing slices in pairs has been part of the implementation from the beginning. @dgursoy suggested (in-person) that this may have been done to more efficiently utilize the FFT operators available 5+ years ago. Nowadays, it may be that libraries have RFFTs, so there may be no performance hit from decoupling adjacent slices.

I'm not sure how trivial it will be to switch the slice indexing to increment by 1 and swap out the "complex" FFT for a real valued FFT. Or maybe slices can continue to be processed in pairs, but `set_filter_tables` should be updated, so it makes a unique table for each slice?
